byterun/compare.c

```diff
@@ -31,61 +31,78 @@
 
 struct compare_item { value * v1, * v2; mlsize_t count; };
 
-#define COMPARE_STACK_INIT_SIZE 256
+#define COMPARE_STACK_INIT_SIZE 8
+#define COMPARE_STACK_MIN_ALLOC_SIZE 32
 #define COMPARE_STACK_MAX_SIZE (1024*1024)
-
-static struct compare_item compare_stack_init[COMPARE_STACK_INIT_SIZE];
-
-static struct compare_item * compare_stack = compare_stack_init;
-static struct compare_item * compare_stack_limit = compare_stack_init
-                                                   + COMPARE_STACK_INIT_SIZE;
-
 CAMLexport int caml_compare_unordered;
 
+struct compare_stack {
+  struct compare_item init_stack[COMPARE_STACK_INIT_SIZE];
+  struct compare_item* stack;
+  struct compare_item* limit;
+};
+
 /* Free the compare stack if needed */
-static void compare_free_stack(void)
+static void compare_free_stack(struct compare_stack* stk)
 {
-  if (compare_stack != compare_stack_init) {
-    free(compare_stack);
-    /* Reinitialize the globals for next time around */
-    compare_stack = compare_stack_init;
-    compare_stack_limit = compare_stack + COMPARE_STACK_INIT_SIZE;
+  if (stk->stack != stk->init_stack) {
+    free(stk->stack);
+    stk->stack = 0;
```

Comment:
If we wanted to avoid an Undefined Behavior if `do_compare_val` or `compare_resize_stack` was called after `compare_free_stack`, then indeed we would need `stk->stack = stk->init_stack`, but if would not be enough, we would also need `stk->stack_limit = stk->init_stack + COMPARE_STACK_INIT_SIZE`. If we don't want to make this safe (the implementation does not currently need it, and it's not clear that it would), then I would agree that dropping the `stk->stack = 0;` write would be better. @stedolan, an explanation for your choice could be interesting. (I believe that the patch could be merged after this is clarified.)